packages/protocol/lib/compatibility/verify-bytecode.ts

```diff
@@ -78,60 +81,63 @@ const getOnchainBytecode = async (address: string, context: VerificationContext)
 const isLibrary = (contract: string, context: VerificationContext) =>
   contract in context.libraryAddresses.addresses
 
-const getImplementationAddress = async (contract: string, context: VerificationContext) => {
-  // Where we find the implementation address depends on two factors:
-  // 1. Is the contract affected by a governance proposal?
-  // 2. Is the contract registered in the Registry or a linked library?
-
-  if (isImplementationChanged(contract, context)) {
-    return getProposedImplementationAddress(contract, context)
-  }
-
-  if (isLibrary(contract, context)) {
-    return `0x${context.libraryAddresses.addresses[contract]}`
-  }
-
-  // contract is registered but we need to check if the proxy is affected by the proposal
-  const proxyAddress = isProxyChanged(contract, context)
-    ? getProposedProxyAddress(contract, context)
-    : await context.registry.getAddressForString(contract)
-
-  // at() returns a promise despite Typescript labelling the await as extraneous
-  const proxy: ProxyInstance = await context.Proxy.at(
-    context.web3.utils.toChecksumAddress(proxyAddress)
-  )
-  return proxy._getImplementation()
-}
-
 const dfsStep = async (queue: string[], visited: Set<string>, context: VerificationContext) => {
   const contract = queue.pop()
   // mark current DFS node as visited
   visited.add(contract)
 
-  // get source code
+  // check proxy deployment
+  if (isProxyChanged(contract, context)) {
+    const proxyAddress = getProposedProxyAddress(contract, context)
+    const governanceAddr = await context.registry.getAddressForString('Governance')
+    // ganache does not support eth_getProof
+    if (
+      context.network !== 'development' &&
+      !(await verifyProxyStorageProof(context.web3, proxyAddress, governanceAddr))
+    ) {
+      throw new Error(`Proposed ${contract}Proxy has impure storage`)
+    }
+
+    const onchainProxyBytecode = await getOnchainBytecode(proxyAddress, context)
+    const sourceProxyBytecode = getSourceBytecode(`${contract}Proxy`, context)
+    if (onchainProxyBytecode !== sourceProxyBytecode) {
+      throw new Error(`Proposed ${contract}Proxy does not match compiled proxy bytecode`)
+    }
+
+    console.log(
+      `Proxy deployed at ${proxyAddress} matches ${contract}Proxy (implementation and storage)`
```

Comment:
Wording nit: I'd change "implementation and storage" to "bytecode and storage", implementation in my mind being reserved for the logic contract pointed to by the Proxy.